numpy/core/src/multiarray/arraytypes.c.src

```diff
@@ -3692,9 +3691,12 @@ static void
         }
     }
     else {
-        for (i = 0; i < ni; i++) {
+        for (i = 0, j = 0; i < ni; i++, j++) {
+            if (j >= nv) {
+                j = 0;
+            }
             if (mask[i]) {
-                in[i] = vals[i%nv];
+                in[i] = vals[i];
```

Comment:
this should be vals[j]
as is the compiler will remove the dead j branch so the benchmarks are not valid, also seems the tests are insufficient.
but I do think it should still be beneficial as the branch should be predicted most of the time